cloudmailer/cloudmailer.py

```diff
@@ -569,92 +556,106 @@ def scheduleReboot(data,hostgroups, hostdict, starttime, interval):
+    """genereate project email text from template"""
+    projmail = ""
+    for line in template:
+        try:
+            if line.find("PROJECT-NAME") != -1:
+                projmail = projmail + project_dict["name"] + "\n"
+                projmail = projmail + "-"*len(project_dict["name"]) + "\n"
+            elif line.find("LIST-OF-MACHINES") != -1:
+                for machine in project_dict["servers"]:
+                    projmail = projmail + machine
+            else:
+                projmail = projmail + line
+        except Exception as e:
+            print ("There was an issue with this project:")
+            pprint.pprint(project_dict)
+            print( e )
+            return False
+    return projmail
+
+def write_copy_of_email_to_file(project_name, emails_str, subject, projmail_str):
+    """Write the project emails to file for review"""
+    file_name = f"{TEMPDIR}/{project_name}"
+    print(f"Creating file '{file_name}' ...")
+    with open(file_name, "w", encoding='UTF-8') as emailcopy:
+        emailcopy.write(f"From: {MAIL_FROM}\n")
+        emailcopy.write(f"To: {emails_str}\n")
+        if MAIL_BCC:
+            emailcopy.write(f"Bcc: {MAIL_BCC}\n")
+        emailcopy.write(f"Subject: {subject}\n")
+        emailcopy.write(projmail_str)
+
+def generate_and_send_emails(send_emails, subject, template, projects):
+    """This function does the following:
+    1. Genereate mail for each project
+    2. Write the generated mails to file
+    3. Send emails to customers if send_emails == True
+    4. Send emails to BCC id send_eamils === True and MAIL_BCC
+        projects{ tenant_id:{ 'name': str(), 'emails': ["example@example.org"] }}
+    """
 
     print(str(len(projects)) + " projects to send email to.")
     ask_for_verification = True
     print(f"Establishing a connection with mail server '{MAIL_SERVER}:25'...")
     smtpconn = smtplib.SMTP(MAIL_SERVER, 25)
     notified_admin = False
 
-    for project in projects.keys():
-        print(f"Processing project '{projects[project]['name']}'...")
-        if len("".join(projects[project]["emails"])) == 0:
-            print("Project %s has no email recipients. PLEASE NOTE that this project will not receive any email!" % projects[project]["name"])
+    for project in projects:
+        print(project)
+        project_name = projects[project]["name"]
+        project_email_address_list = projects[project]['emails']
+        print(f"Processing project '{project_name}'...")
+        if len(project_email_address_list) == 0:
+            print(f"Project {project_name} has no email recipients. PLEASE NOTE that this " \
+                  "project will not receive any email!")
             continue
+        projmail_str = generate_mail_text(projects[project], template)
+        if not projmail_str:
+            send_emails = False
+            print( "Emails won't be sent for {project_name} because of exception")
+            continue
```

Comment:
This if statement seems to be a bad idea but it was there previously as well. IMO this means that projects after his project won't receive any emails. IMO it might better to remove `send_emails = False` .

Reply:
You are right in your interpretation of it as a bug, i.e., it will affect the projects handled after the current one too.
In your proposed new version the `continue` statement allows to simply skip to the next project, so effectively you do not need that line anymore.

Reply:
Do you want me to remove the `continue` and keep the `send_emails = False` ?

Reply:
I think the other way around. just keep the continue? That's the correct way right?